core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -233,17 +235,13 @@ private[spark] class MemoryStore(
     }
 
     if (keepUnrolling) {
-      // We successfully unrolled the entirety of this block
-      val arrayValues = vector.toArray
-      vector = null
-      val entry =
-        new DeserializedMemoryEntry[T](arrayValues, SizeEstimator.estimate(arrayValues), classTag)
-      val size = entry.size
+      // We need more precise value
+      val size = valuesHolder.esitimatedSize(false)
```

Comment:
why do we need `esitimatedSize(false)`? It seems we can just build the entry and call `entry.size`.

Reply:
It can avoid creating an useless entry if the condition `enoughStorageMemory  == true` does not satisfy, I guess.

Reply:
Looking at the implementations, I don't think this makes a lot of difference. `esitimatedSize(false)` already does some heavy work and makes this `ValuesHolder` unusable, i.e. you can't call `storeValue` anymore, which seems not something a size estimation should do.

Reply:
`roughly = false` tells more than estimating the size of vector, which is 'unroll has finished'. So, `storeValue` will not be called anymore.

And for `heavy work`(under my understanding, mostly due to `SizeEstimator.estimate(arrayValues)`), yeah, I agree with you. Those `heavy work` would be done whether we create an entry or not or whether the entry put into the memory store or not finally.

Reply:
I change the code back to originally. For `DeserializedValuesHolder`, we could `buildEntry` and get the `size` from `MemorySize`. But for `SerializedValuesHolder`, this way not work correctly. Because we need call the `bbos.toChunkedByteBuffer` to get the `MemoryEntry` object, and if the reserved memory is not enough for transfer the unroll memory to storage memory. Then we unroll failed and need call `bbos.toChunkedByteBuffer` ([L802](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala#L802), this should be intentional which related to #15043). So the problem is that we call `bbos.toChunkedByteBuffer` twice, but it can't be.